### DISCUSSION TITLE
Logging tools

### DISCUSSION THREAD

**Discussion by simonasmulevicius** [microsoft/msquic#1161] [Q&A] [answered]
1. Is _msquic_ able to output logs in the `.qlog` format? I would like to use _qvis_ (https://qvis.edm.uhasselt.be/#/files) to visualise and understand the way _msquic_ works.
2. What other logging tools are you using to capture and analyse QUIC packets (e.g. some integration with Wireshark)?

Thanks!

**Comment by nibanks** [comment, marked as answer]
The logging story for MsQuic is quite complex. We strive for extremely high performant logging (i.e. no/minimal overhead at gigabit transfer rates) so we cannot use qlog directly. Instead, on Windows we use manifested ETW and on Linux we use LTTng binary logging formats. Additionally, I created a very minimal ETW -> qlog convertor, but it's now woefully out of date.

That being said, we do have significant tools of our own to analyze ETW traces (even in more detail in some aspects than qvis). We use Windows Performance Analyzer (WPA). All the necessary bits are currently going through the final approval stages to be shared externally to Microsoft (ETA this month). I will update this when I have more exact information.

**Reply by nibanks** [reply]
Just created a PR with some initial instructions https://github.com/microsoft/msquic/pull/1167.

**Reply by nibanks** [reply]
I have merged that PR and updated the [Diagnostics docs](https://github.com/microsoft/msquic/blob/main/docs/Diagnostics.md) for the latest information, which includes a link to the WPA instructions (which are still a WIP). If you have specific questions or requests for additional information, please let me know.

**Reply by nibanks** [reply]
And one more thing, we do have an issue tracking official qlog support: https://github.com/microsoft/msquic/issues/1158
